wallet: Implement independent BDB parser

achow101 opened the pull request:
Split from #26596 

This PR adds `BerkeleyRODatabase` which is an independent implementation of a BDB file parser. It provides read only access to a BDB file, and can therefore be used as a read only database backend for wallets. This will be used for dumping legacy wallet records and migrating legacy wallets without the need for BDB itself.

Wallettool's `dump` command is changed to use `BerkeleyRODatabase` instead of `BerkeleyDatabase` (and `CWallet` itself) to demonstrate that this parser works and to test it against the existing wallettool functional tests.

Marek777777 reviewed (changes requested):
.

Sjors commented on src/streams.h:
```diff
@@ -530,6 +530,14 @@ class AutoFile
     //
+    void seek(int64_t offset, int origin)
+    {
+        if (!file)
```
1d093bbf97b8d4ec5413ca560ac79e56f23d7f1d `{`

Could also use the `IsNull()` helper.

achow101 replied:
Done

Sjors commented on src/streams.h:
```diff
@@ -530,6 +530,14 @@ class AutoFile
     //
     // Stream subset
     //
+    void seek(int64_t offset, int origin)
```
1d093bbf97b8d4ec5413ca560ac79e56f23d7f1d: Note for other reviewers: yes

https://en.cppreference.com/w/c/io/fseek

Sjors commented on src/streams.h:
```diff
@@ -530,6 +530,14 @@ class AutoFile
+        if (!file)
+            throw std::ios_base::failure("CAutoFile::read: file handle is nullptr");
+        if (fseek(file, offset, origin) != 0)
+            throw std::ios_base::failure(feof(file) ? "CAutoFile::seek: end of file" : "CAutoFile::seek: fseek failed");
```
1d093bbf97b8d4ec5413ca560ac79e56f23d7f1d: maybe log the exact error code.

achow101 replied:
It's kind of annoying to do that, so I'll leave it as is. The other functions do the same.

Sjors commented on src/wallet/migrate.h:
```diff
@@ -0,0 +1,104 @@
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_WALLET_MIGRATE_H
+#define BITCOIN_WALLET_MIGRATE_H
```
1224f45a92972199799d1ea33700eee567dc66fe: maybe call this file `bdb_ro_wallet_db.h`?

achow101 replied:
The intent is to have all of the migration stuff in these files, so I will leave it as is.

Sjors commented on src/wallet/migrate.cpp:
```diff
@@ -21,17 +21,37 @@ bool BerkeleyRODatabase::Backup(const std::string& dest) const
 
 bool BerkeleyROBatch::ReadKey(DataStream&& key, DataStream& value)
 {
-    return false;
```
d6c5aa0f6783ef9f26f93a583d2fad5a943e23ec: commit description also mentions `StartCursor`, `ReadAtCursor`, `CloseCursor`.

Sjors reviewed (commented):
Some review nibbles.

achow101 commented:
> Have you looked into a fuzz target for this parser?

I haven't. If someone would like to write one, I'd be happy to add it.

Sjors commented:
Adding a fuzz target that compares this to the real thing seems useful. Given that this PR changes `dumpwallet` to rely on the new parser, I think we should do that sooner rather than later. Otherwise I suggest changing e88ad38ac2f3e43de54a332965b13eb7cc27b91f so that the read-only parser is optional, and then change the tests to always use it.

It seems a bit daunting to compare the low level DBD parsing code in this PR to the real thing. Having a fuzzer also reduces the need to be very thorough in that. I'm not sure how much confidence I have in the existing test coverage for dumpwallet.

TheCharlatan commented:
Concept ACK

sedited reviewed (commented):
Approach ACK

I'm testing this against my hand-rolled databases, so far this looks solid.

sedited committed:
Berkeley RO Database fuzz test

TheCharlatan commented:
Wrote [my first fuzz test](https://github.com/TheCharlatan/bitcoin/commit/995ccef131e71ae94f5c73ee3b0297282f5646ed) (I really don't know what I am doing so not even sure this is useful), and managed to hit: `fuzz: wallet/migrate.cpp:384: void wallet::RecordsPage::Unserialize(Stream &) [Stream = CAutoFile]: Assertion false failed`.

achow101 commented:
> Wrote [my first fuzz test](https://github.com/TheCharlatan/bitcoin/commit/995ccef131e71ae94f5c73ee3b0297282f5646ed) (I really don't know what I am doing so not even sure this is useful), and managed to hit: `fuzz: wallet/migrate.cpp:384: void wallet::RecordsPage::Unserialize(Stream &) [Stream = CAutoFile]: Assertion false failed`.

Do you have the input that hit that?

TheCharlatan commented:
Re https://github.com/bitcoin/bitcoin/pull/26606#issuecomment-1601156957

> Do you have the input that hit that?

Yes, pasting as base64 here, since I'm not sure about the practice of attaching random binary files:

```
echo "HQABAAAAAQAAAACPAAUxYgAAAAmK/////wkAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAKysrAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA/////gAAAAAAz8/Pz8/Pz8/Pz8/Pz8/Pz8/Pz8/Pz8/Pz8/Pz8/Pz8/Pz8/Pz8/Pz8/Pz8/Pz8/Pz8/Pz8/Pz8/Pz8/Xz8/Pz8/Pz8/Pz8/Pz8/Pz88Az8/Pz8/PADqKIIr/////////EwAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAArKysAAAAAAAAAAAAAAAAAAADPz8/Pz8/PxXt7e3t7e3t7AABPAAUxz8/Pz8/Pz8/Pz8/Pz8/Pz8/Pz8/Pz8/Pz8/Pz8/Pz8/S0tLS0tLS0tLS0tLS0tLS0tLS0tLS0nh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eNLS0tLS0tLSKtLS0tLS0tLS0tLS0tLS0tDS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0tLS0s/Pz8/Pz8/Pz8/Pz8/PAAB7e3t7e3t7e3sBAAABAAACK08ABQAgXFxcXFxcAQAAAACPAAUxYgAAAAmK/////////////wAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAKysrAAAAAAAAAAAAAAAAAAAAz8/Pz8/Pz8/Pz8/Pz8/Pz8/Pz8/Pz8/PzwAAAP/Pz8/Pz8/Pz8/Pz8/Pz8/Pz8/Pz8/Pz8/Pz8/Pz8/Pzw==" | base64 -d > crash-assert
```

achow101 commented:
Found an issue with prefix cursor handling, updated the tests to check for that. Also pulled in @TheCharlatan's fuzz test and changed several asserts to instead throw exceptions which resolves the fuzz crash.

achow101 commented on src/streams.h:
```diff
@@ -522,6 +522,16 @@ class AutoFile
+    void seek(int64_t offset, int origin)
+    {
+        if (IsNull()) {
+            throw std::ios_base::failure("CAutoFile::read: file handle is nullptr");
```
Done

maflcko reviewed (commented):
(nit)

TheCharlatan commented on src/wallet/test/fuzz/wallet_bdb_parser.cpp:
```diff
@@ -0,0 +1,79 @@
+using wallet::DatabaseStatus;
+
+namespace {
+const TestingSetup* g_setup;
```
After testing with a valid database in the fuzzing corpus, I noticed that I messed up the `m_args`, `m_node.args`, `gArgs` division again ([see here](https://github.com/bitcoin/bitcoin/issues/25055)) - sorry about that. I have a basic patch here, but feel free to commit a more elegant fix:

```diff
diff --git a/src/wallet/test/fuzz/wallet_bdb_parser.cpp b/src/wallet/test/fuzz/wallet_bdb_parser.cpp
index aabde093b6..a1acb95178 100644
--- a/src/wallet/test/fuzz/wallet_bdb_parser.cpp
+++ b/src/wallet/test/fuzz/wallet_bdb_parser.cpp
@@ -19,23 +19,15 @@ using wallet::DatabaseOptions;
 using wallet::DatabaseStatus;
 
 namespace {
-const TestingSetup* g_setup;
+TestingSetup* g_setup;
 } // namespace
 
 void initialize_wallet_bdb_parser()
 {
-    static const auto testing_setup = MakeNoLogFileContext<const TestingSetup>();
+    static auto testing_setup = MakeNoLogFileContext<TestingSetup>();
     g_setup = testing_setup.get();
 }
 
-void SetDumpFile(ArgsManager& args) {
-    auto dumpfile{args.GetDataDirNet() / "fuzzed_dumpfile.dat"};
-    if (fs::exists(dumpfile)) { // Writing into an existing dump file will throw an exception
-        remove(dumpfile);
-    }
-    args.ForceSetArg("-dumpfile", fs::PathToString(args.GetDataDirNet() / "fuzzed_dumpfile.dat"));
-}
-
 FUZZ_TARGET_INIT(wallet_bdb_parser, initialize_wallet_bdb_parser)
 {
     FuzzedDataProvider fuzzed_data_provider{buffer.data(), buffer.size()};
@@ -52,10 +44,14 @@ FUZZ_TARGET_INIT(wallet_bdb_parser, initialize_wallet_bdb_parser)
     DatabaseStatus status;
     bilingual_str error;
 
+    auto dumpfile{g_setup->m_args.GetDataDirNet() / "fuzzed_dumpfile.dat"};
+    if (fs::exists(dumpfile)) { // Writing into an existing dump file will throw an exception
+        remove(dumpfile);
+    }
+    g_setup->m_args.ForceSetArg("-dumpfile", dumpfile);
+
     try {
         auto db{MakeBerkeleyRODatabase(wallet_path, options, status, error)};
-        const auto& node = g_setup->m_node;
-        SetDumpFile(*node.args);
         assert(DumpWallet(g_setup->m_args, *db, error));
     }
     catch (const std::runtime_error& e) {
```

achow101 replied:
Done as suggested

TheCharlatan commented on src/wallet/migrate.cpp:
```diff
@@ -519,7 +582,7 @@ void BerkeleyRODatabase::Open()
+                        PageHeader opage_header(inner_meta.other_endian);
                         db_file >> opage_header;
                         OverflowPage opage(opage_header);
                         db_file >> opage;
```
In commit "Implement handling of other endianness in BerkeleyRODatabase" 0b8eac6a78351a68c1b5a6126564493ab50031dc

I might be reading something wrong, or confusing something, but I don't quite understand why the record data itself does not have to get its endianness adjusted.

achow101 replied:
The record data should just be arbitrary byte strings so they don't have an endianness.

TheCharlatan replied:
ACK, please resolve.

TheCharlatan commented on src/wallet/migrate.cpp:
```diff
@@ -1,13 +1,547 @@
+
+    static constexpr size_t SIZE = 9; // Overflow record is always 9 bytes
+
+    std::vector<std::byte> data; // Data from all of the overflow pages
```
In commit "wallet: implement independent BDB deserializer in BerkeleyRODatabase" d5fe130106bdf367c0bca157b02d12254590b585

Can this `data` field be removed? Seems to be unused.

achow101 replied:
Removed. I think originally I though I could put the overflow page data into this field, but I didn't end up doing it that way.

achow101 commented on src/streams.h:
```diff
@@ -522,6 +522,16 @@ class AutoFile
+    void seek(int64_t offset, int origin)
+    {
+        if (IsNull()) {
+            throw std::ios_base::failure("AutoFile::read: file handle is nullptr");
```
Done

achow101 commented on src/wallet/migrate.h:
```diff
@@ -0,0 +1,122 @@
+    const BerkeleyRODatabase& m_database;
+
+    bool ReadKey(DataStream&& key, DataStream& value) override;
+    bool WriteKey(DataStream&& key, DataStream&& value, bool overwrite=true) override { return true; }
```
IIRC I did this to avoid issues with old wallets that the wallet loading logic would do automatic upgrading of which expects writing to not fail.

josibake replied:
nit: might be worth adding a comment here explaining this. I was also surprised to see this not returning `false` and couldn't understand why this was different from other methods until I read your comment here.

achow101 replied:
Added a comment.

TheCharlatan commented on src/wallet/migrate.cpp:
```diff
@@ -0,0 +1,683 @@
+    db_file >> page;
+
+    // First record should be the string "main"
+    if (!std::holds_alternative<DataRecord>(page.records.at(0)) || std::get_if<DataRecord>(&page.records.at(0))->data != std::vector<std::byte>({std::byte('m'), std::byte('a'), std::byte('i'), std::byte('n')})) {
```
This would be more readable if the `main` bytes were stuffed into a constant somewhere.

achow101 replied:
Done

TheCharlatan commented on src/wallet/migrate.cpp:
```diff
@@ -0,0 +1,683 @@
+    if (!std::holds_alternative<DataRecord>(page.records.at(0)) || std::get_if<DataRecord>(&page.records.at(0))->data != std::vector<std::byte>({std::byte('m'), std::byte('a'), std::byte('i'), std::byte('n')})) {
+        throw std::runtime_error("Subdatabase has an unexpected name");
+    }
+    if (!std::holds_alternative<DataRecord>(page.records.at(1)) || std::get_if<DataRecord>(&page.records.at(1))->m_header.len != 4) {
```
Can you add a comment explaining why we check this?

achow101 replied:
Added a comment

sedited reviewed (commented):
ACK 6e28ff97a99519ec8b50123bc1177084bba68f96 (besides the two small comments)

I'm sure I could spend many hours more verifying the parser, but at this point I'd like to encourage other reviewers to take a look. I think if this is merged soon and given the opportunity to be fuzzed against a large corpus of both random seeds, bdb databases, and valid wallets we could also gain some confidence in the code.

Sjors commented:
Concept ACK

I still suggest changing fc810c1df899e5b42fe7658ebdddc6b4320240a6 so that `dumpwallet` does not use the read-only format by default, and instead change the tests use it. However with the new fuzz target commit I'm less worried about it.

Sjors commented on src/wallet/test/fuzz/wallet_bdb_parser.cpp:
```diff
@@ -0,0 +1,75 @@
+{
+    FuzzedDataProvider fuzzed_data_provider{buffer.data(), buffer.size()};
+
+    const auto wallet_path = g_setup->m_args.GetDataDirNet() / "fuzzed_wallet.dat";
```
Maybe add a recommendation to the fuzz doc to [mount /tmp to tmpfs](https://ubuntu.com/blog/data-driven-analysis-tmp-on-tmpfs), or ensure we don't use the file system.

Sjors replied:
Not in the fuzz docs, but just having it in the test docs is probably fine.

Sjors commented on src/wallet/test/fuzz/wallet_bdb_parser.cpp:
```diff
@@ -0,0 +1,75 @@
+
+    try {
+        auto db{MakeBerkeleyRODatabase(wallet_path, options, status, error)};
+        assert(DumpWallet(g_setup->m_args, *db, error));
```
It would be useful here - if BDB is compiled - to also open it with BDB, dump and then compare both dumps.

I'm not sure what to do for cases where BDB and RO have slightly different opinions on what they're willing to open at all. It's probably fine to skip comparing files where _either_ refuses to open.

achow101 replied:
Added a comparison

Sjors replied:
Nice! I'll run the fuzzer against that...

achow101 commented:
> I still suggest changing fc810c1 so that dumpwallet does not use the read-only format by default, and instead change the tests use it.

Since dumping should be a read-only operation, I think it's better to use the read-only database since it definitely won't write to it like BDB does. Also it provides a good mechanism for testing.

maflcko reviewed (commented):
Looks like CI is red?

Sjors commented:
> Since dumping should be a read-only operation, I think it's better to use the read-only database since it definitely won't write to it like BDB does.

But it could produce a corrupt dump, which someone might rely on for their backup (even if they shouldn't).

Sjors commented on src/wallet/test/fuzz/wallet_bdb_parser.cpp:
```diff
@@ -0,0 +1,108 @@
+        if (std::string(e.what()) == "Unexpected page size") return;
+        if (std::string(e.what()) == "Unexpected page type") return;
+        throw e;
+    }
```
So far my corpus doesn't make it here, as tested by sprinkling `assert(false)`. I may indeed need some help with a real wallet.

Sjors commented on src/wallet/test/fuzz/wallet_bdb_parser.cpp:
```diff
@@ -0,0 +1,108 @@
+
+    {
+        AutoFile outfile{fsbridge::fopen(wallet_path, "wb")};
+        const auto file_data{ConsumeRandomLengthByteVector(fuzzed_data_provider)};
```
I simply copied a `wallet.dat` into the corpus directory. That resulted in a bunch of extra corpus entities and almost immediately a crash.

Sjors replied:
Though the file seems to get ignored when trying to merge.

Sjors replied:
Ah, figured it out! Will push some new entries to the qa-assets PR that should reproduce the crash.

achow101 commented on src/streams.h:
```diff
@@ -528,6 +528,7 @@ class AutoFile
     //
     // Stream subset
     //
+    void seek(int64_t offset, int origin);
```
Done

Sjors commented:
I tried adding `if (bdb_ro_err) return;` above `#ifdef USE_BDB` (although that does reduce coverage). I also added in the second catch:

```cpp
// We are not interested in fuzzing the original BDB implementation
if (std::string(e.what()) == "BerkeleyDatabase: Error 22, can't open database fuzzed_wallet.dat") return;
```

That helps a bit, but still leads to a "SEGV on unknown address 0x000000000000" in `DumpWallet` very quickly. I haven't looked at what the fuzzer came up with. Perhaps it can be dealt with by making the RO parser more strict?

Or can we tell the fuzzer to ignore crashes in the original BDB entirely? We only care about comparing dumps for wallets that the original implementation can load.

achow101 commented:
I've fixed at least one fuzzer crash. BDB verifies a couple of other things that RO was not.

I also changed the fuzzer to make sure that BDB fails to open the database if RO failed to open it, and vice versa.

There's still another crash that I have, but it's something where RO is being stricter than BDB.

Sjors commented:
> There's still another crash that I have

Not sure if that was fixed in your later push, but in any case, I pushed another crash to the qa-assets PR.

maflcko reviewed (commented):
(nits only)

achow101 commented:
> But it could produce a corrupt dump, which someone might rely on for their backup (even if they shouldn't).

I've changed it to be optional and enabled with the argument `-withinternalbdb`.

> Not sure if that was fixed in your later push, but in any case,

Yes, I fixed that

> I pushed another crash to the qa-assets PR.

This seems to be crash in BDB itself rather than in our parser.

Sjors commented:
> Not entirely sure how to handle that since we don't care about fuzzing BDB itself.

Indeed, @MarcoFalke any idea on how to make the fuzzer ignore that?

theStack reviewed (commented):
Concept ACK

As a follow-up, it might be nice if we could somehow add coverage for the "other endianness" code path in the tests. Did that manually with the following patch (creates the BDB database in big-endian format, while I'm running a little-endian system):
```diff
diff --git a/src/wallet/bdb.cpp b/src/wallet/bdb.cpp
index 9ea43ca67c..945e12ebdb 100644
--- a/src/wallet/bdb.cpp
+++ b/src/wallet/bdb.cpp
@@ -389,6 +389,7 @@ void BerkeleyDatabase::Open()
                 }
             }
 
+            pdb_temp->set_lorder(4321);
             ret = pdb_temp->open(nullptr,                             // Txn pointer
                             fMockDb ? nullptr : strFile.c_str(),      // Filename
                             fMockDb ? strFile.c_str() : "main",       // Logical db name
```
Tests still passed, only had to adapt the BDB magic number check in wallet_migration.py which currently assumes LE byte order.

TheCharlatan commented:
Re https://github.com/bitcoin/bitcoin/pull/26606#issuecomment-1669359482

> What are the steps to reproduce?

Run the fuzzer on the fuzz test introduced in this PR seeded with Sjors' inputs from his [qa repo PR](https://github.com/bitcoin-core/qa-assets/pull/140) and compiled with bdb enabled.

> Outside of that, my recommendations would be to use bdb-5.3

The bugs persist even with bdb 5.3.

achow101 commented on src/wallet/migrate.cpp:
```diff
@@ -21,17 +21,41 @@ bool BerkeleyRODatabase::Backup(const std::string& dest) const
+    if (m_database.m_records.count(key_data) == 0) {
+        return false;
+    }
+    auto val = m_database.m_records.at(key_data);
```
Done as suggested

achow101 commented on src/wallet/migrate.cpp:
```diff
@@ -1,13 +1,601 @@
+    db_file >> page;
+
+    // First record should be the string "main"
+    if (!std::holds_alternative<DataRecord>(page.records.at(0)) || std::get_if<DataRecord>(&page.records.at(0))->data != SUBDATABASE_NAME) {
```
Done as suggested

sedited reviewed (approved):
Re-ACK 3c91202c5a03c57c9f83186de9d3caf337b78d45

achow101 committed:
Add AutoFile::seek and tell

It's useful to be able to seek to a specific position in a file. Allow
AutoFile to seek by using fseek.

It's also useful to be able to get the current position in a file.
Allow AutoFile to tell by using ftell.

achow101 committed:
wallet: add dummy BerkeleyRODatabase and BerkeleyROBatch classes

BerkeleyRODatabase and BerkeleyROBatch will be used to access a BDB file
without the use of BDB. For now, these are dummy classes.

achow101 committed:
wallet: implement BerkeleyROBatch

Implement ReadKey and HasKey of BerkeleyROBatch, and Next of BerkeleyROCursor.
Also adds the containers for records to BerkeleyRODatabase so that
BerkeleyROBatch will be able to access the records.

achow101 committed:
wallet: implement BerkeleyRODatabase::Backup

achow101 committed:
wallet: implement independent BDB deserializer in BerkeleyRODatabase

BerkeleyRODatabase is intended for use after BDB is removed, so it needs
to be able to read all of the records from a BDB file. Thus an
independent deserializer for BDB data files is implemented in it. This
deserializer is targeted towards the data files that Bitcoin Core
creates so it does not fully support all of BDB's features (e.g. other
database types, encryption, etc.).

achow101 committed:
Implement handling of other endianness in BerkeleyRODatabase

achow101 committed:
Add MakeBerkeleyRODatabase

Implements MakeBerkeleyRODatabase and adds DatabaseFormat::BERKELEY_RO
so that MakeDatabase can use BerkeleyRO as the backend database.

achow101 committed:
wallettool: Optionally use BERKELEY_RO as format when dumping BDB wallets

In order to ease the transition to not having BDB, make the dump tool
use DatabaseFormmat::BERKELEY_RO when -withinternalbdb is set.

achow101 committed:
tests: Add BerkeleyRO to db prefix tests

achow101 commented on src/wallet/walletdb.cpp:
```diff
@@ -1476,6 +1477,11 @@ std::unique_ptr<WalletDatabase> MakeDatabase(const fs::path& path, const Databas
         return nullptr;
     }
 
+    // If BERKELEY_RO was the format, then change the format from BERKELEY to BERKELEY_RO
```
Indeed, fixed.

achow101 commented:
Rebased for silent merge conflict

laanwj commented on src/wallet/migrate.cpp:
```diff
@@ -0,0 +1,738 @@
+
+namespace wallet {
+// Magic bytes in both endianness's
+constexpr uint32_t BTREE_MAGIC = 0x00053162; // If the file endianness matches our system, we see this magic
```
Just a note for other reviewers: This comment mentioning "our system" was slightly confusing to me: our serialization framework always serializes in LE, so if we see BTREE_MAGIC_OE it means the file is in big-endian. Which means we always need to swap it, independent of system's endianness.

laanwj commented on src/wallet/migrate.cpp:
```diff
@@ -0,0 +1,738 @@
+    uint32_t key_count;       // Cached key count
+    uint32_t record_count;    // Cached record count
+    BTreeFlags flags;         // Flags
+    uint160 uid;              // Unique file ID (20 bytes, fits in uint16)
```
i don't think importing `uint160` is necessary here--the actual value isnt used, we're just using it to consume some bytes

achow101 replied:
Dropped `uint160`, changed to use an array.

laanwj commented:
Reference for reviewers:
- `db_page.h` from berkeleydb 4.8.30: https://gist.github.com/laanwj/2b998d6db8d9df384a30871035675aec

murchandamus commented on src/wallet/migrate.cpp:
```diff
@@ -1,13 +1,601 @@
+
+static void SeekToPage(AutoFile& s, uint32_t page_num, uint32_t page_size)
+{
+    size_t pos = page_num * page_size;
```
calculation of two 32 bit integers, you need to use a cast

laanwj replied:
aside: might be better to not use size_t at all for file offsets, as the type is 32 bit on usual 32 bit platforms

laanwj commented on src/wallet/migrate.cpp:
```diff
@@ -0,0 +1,738 @@
+        s >> magic;
+        s >> version;
+        s >> pagesize;
+        s >> encrypt_algo;
```
do we want to sanity check for encryption to be disabled? (didn't know bdb supported db-level encryption in the first place)

achow101 replied:
Yes, I've added a check for that.

laanwj commented:
i tried some old wallets, and there was a problem with parsing the headers of overflow pages (which have btree level 0), the following makes the check pass and adds a further check that overflow records correspond to overflow pages:
```patch
diff --git a/src/wallet/migrate.cpp b/src/wallet/migrate.cpp
index be54b58f67c9bc961f8df306fe4bc0bf810b41f3..1c4ce1c1ed8d44221626613df2cd0d4712fc5c85 100644
--- a/src/wallet/migrate.cpp
+++ b/src/wallet/migrate.cpp
@@ -361,7 +361,7 @@ public:
         if (expected_page_num != page_num) {
             throw std::runtime_error("Page number mismatch");
         }
-        if (level < 1) {
+        if ((type != PageType::OVERFLOW_DATA && level < 1) || (type == PageType::OVERFLOW_DATA && level != 0)) {
             throw std::runtime_error("Bad btree level");
         }
     }
@@ -636,6 +636,9 @@ void BerkeleyRODatabase::Open()
                         SeekToPage(db_file, next_page, page_size);
                         PageHeader opage_header(next_page, inner_meta.other_endian);
                         db_file >> opage_header;
+                        if (opage_header.type != PageType::OVERFLOW_DATA) {
+                            throw std::runtime_error("Bad overflow record page type");
+                        }
                         OverflowPage opage(opage_header);
                         db_file >> opage;
                         data.insert(data.end(), opage.data.begin(), opage.data.end());
```

achow101 commented:
> parsing the headers of overflow pages (which have btree level 0)

Interesting, did not know that they would use an invalid btree level. Fixed as suggested.

***

I've also commented out the enums that we do not use. I decided to leave them there for completeness, but they won't be used.

***

For help with reviewing, I suggest looking at the BDB source code. I've uploaded a copy of the version that we use at https://github.com/achow101/bdb-sources/tree/main/4.8.30.NC. The files that I used for referencing and checking that things were implemented correctly are [`dbinc/db_page.h`](https://github.com/achow101/bdb-sources/blob/main/4.8.30.NC/dbinc/db_page.h), [`btree/bt_open.c`](https://github.com/achow101/bdb-sources/blob/main/4.8.30.NC/btree/bt_open.c), and [`btree/bt_put.c`](https://github.com/achow101/bdb-sources/blob/main/4.8.30.NC/btree/bt_put.c). That repo also has a couple of other BDB versions that people may have used for their wallets (except 18.x), so you may want to compare with those versions and check that the file format did not change between them.

achow101 committed:
Error if LSNs are not reset

achow101 commented:
There was a comment at CoreDev that this should check that the file is compacted, so I've added a commit that does that. To check that this is correct, see the implementation of [`lsn_reset`](https://github.com/achow101/bdb-sources/blob/main/4.8.30.NC/db/db_setlsn.c#L114), which uses [`LSN_NOT_LOGGED`](https://github.com/achow101/bdb-sources/blob/main/4.8.30.NC/dbinc/db_int.in#L803). The definition of [`DB_LSN`](https://github.com/achow101/bdb-sources/blob/main/4.8.30.NC/dbinc/db.in#L444) is also useful.

laanwj commented:
> There was a comment at CoreDev that this should check that the file is compacted, so I've added a commit that does that.

That's good to have. It does look like the newly added check trips up in the CI in some places:
```
unknown location(0): fatal error: in "db_tests/db_cursor_prefix_range_test": class std::runtime_error: LSNs are not reset, this database is not completely flushed. Please reopen then close the database with a version that has BDB support
```

achow101 commented:
> That's good to have. It does look like the newly added check trips up in the CI in some places:
> 
> ```
> unknown location(0): fatal error: in "db_tests/db_cursor_prefix_range_test": class std::runtime_error: LSNs are not reset, this database is not completely flushed. Please reopen then close the database with a version that has BDB support
> ```

Fixed

laanwj commented on src/wallet/migrate.cpp:
```diff
@@ -0,0 +1,771 @@
+        PageHeader header(curr_page, inner_meta.other_endian);
+        db_file >> header;
+        if (header.type != PageType::BTREE_LEAF && header.type != PageType::BTREE_INTERNAL) {
+            throw std::runtime_error("Unexpected inner database page type");
```
`header.type` is compared against `BTREE_LEAF` and `BTREE_INTERNAL`, which are the same cases as handled in the `switch()` statement below, which also errors in the default page (though with a slightly different error), so this check is redundant?

achow101 replied:
Good point. Removed this check.

laanwj commented on src/wallet/migrate.cpp:
```diff
@@ -0,0 +1,771 @@
+
+            // Go to the offset from the index
+            int64_t to_jump = index - pos;
+            s.ignore(to_jump);
```
Note that the argument to `ignore()` is unsigned (a `size_t`), might want to error out here if somehow `index < pos`.
(same for the occurence in `InternalPage`)

achow101 replied:
Added a check

laanwj commented on src/wallet/migrate.cpp:
```diff
@@ -0,0 +1,771 @@
+    FILE *file = fsbridge::fopen(m_filepath, "rb");
+    AutoFile db_file(file);
+    if (db_file.IsNull()) {
+        db_file.fclose();
```
It looks like `fclose()` does nothing if `db_file.IsNull()` already holds.

achow101 replied:
Removed

laanwj commented on src/wallet/migrate.cpp:
```diff
@@ -0,0 +1,771 @@
+                    key.clear();
+                }
+                is_key = !is_key;
+            }
```
Could add a check that `is_key` is true at the end. If not, there's an odd number of values, which would be odd?

achow101 replied:
Good point. I've added a check earlier that the number of records is even to avoid this kind of issue.

laanwj commented on src/wallet/migrate.cpp:
```diff
@@ -0,0 +1,771 @@
+
+    RecordType GetRealType() const
+    {
+        return static_cast<RecordType>(static_cast<uint8_t>(type) & ~static_cast<uint8_t>(RecordType::DELETE));
```
Instead of casting back and forth and having a `GetRealType()` it would, imo, be more straightforward to split the delete flag off from the type in `Unserialize` before casting to `RecordType`. So to have
```
RecordType type;
bool deleted;
```
Then these accessors could go.

achow101 replied:
Done as suggested.

laanwj commented:
ACK f4f7eda0bf0e798d305df331c1d94443b81dff22

i've extensively reviewed the code and tested `bitcoin-wallet  -withinternalbdb dump ...` on a fair number of historical wallet database backups, getting the exact same dump as BerkeleyDB, or in one case the "LSNs are not reset" error (because it wasn't cleanly closed at the time).

There are a lot of checks and i'm reasonably confident that this won't silently drop or corrupt any data (in one way safer than BDB: it truly opens read-only), instead fails when it sees something is outside expectations. It's of course possible some legit wallet will trigger an edge case exception, but that will only become evident after this reaches wider testing. As this doesn't remove the BerkeleyDB dependency yet, it's not something to be excessively worried about for this PR.

laanwj commented on src/wallet/migrate.cpp:
```diff
@@ -557,6 +557,20 @@ void BerkeleyRODatabase::Open()
+        uint32_t file;
+        uint32_t offset;
+        SeekToPage(db_file, i, page_size);
+        db_file >> file >> offset;
```
The check currently always fails for other-endian wallets, needs something like:
```diff
diff --git a/src/wallet/migrate.cpp b/src/wallet/migrate.cpp
index 15478e8bc5bfe39e04b400fa475215e25eef6452..0a41ae199e1950bf372bae90b8a88fccca2b88f0 100644
--- a/src/wallet/migrate.cpp
+++ b/src/wallet/migrate.cpp
@@ -566,6 +566,10 @@ void BerkeleyRODatabase::Open()
         uint32_t offset;
         SeekToPage(db_file, i, page_size);
         db_file >> file >> offset;
+        if (outer_meta.other_endian) {
+            file = internal_bswap_32(file);
+            offset = internal_bswap_32(offset);
+        }
         if (file != 0 || offset != 1) {
             throw std::runtime_error("LSNs are not reset, this database is not completely flushed. Please reopen then close the database with a version that has BDB support");
         }
```

achow101 replied:
Good catch, fixed

achow101 committed:
bdb: Be able to make byteswapped databases

Byteswapped databases make it easier to test opening and deserializing
other endian databases.

achow101 committed:
test: Test dumps of other endian BDB files

achow101 committed:
test: Test dumping dbs with overflow pages

achow101 commented:
For testing big endian wallets, you can use `db_dump wallet.dat | db_load -c db_lorder=4321 wallet_big_endian.dat` to make big endian databasese using BDB's tools.

However, I also realized that BDB lets us set the endianness to use for a new database, so I've added a `bdb_swap` format to `createfromdump` that will make a BDB database with the other endianness. This will flip it depending on the system that you use, so on little endian, it will make a big endian wallet, and on big endian, it will make a little endian wallet. Adding this lets us have a test of `BerkeleyRO` dumping other endian wallets.

I've also added a test for the overflow page issue that was discovered earlier.

achow101 commented on src/wallet/migrate.cpp:
```diff
@@ -16,7 +17,25 @@ std::unique_ptr<DatabaseBatch> BerkeleyRODatabase::MakeBatch(bool flush_on_close
 bool BerkeleyRODatabase::Backup(const std::string& dest) const
 {
-    return false;
+    fs::path src(m_filepath);
+    fs::path dst(fs::PathFromString(dest));
+
+    if (fs::is_directory(dst)) {
+        dst = BDBDataFile(dst);
+    }
+    try {
+        if (fs::exists(dst) && fs::equivalent(src, dst)) {
+            LogPrintf("cannot backup to wallet source file %s\n", fs::PathToString(dst));
+            return false;
+        }
+
+        fs::copy_file(src, dst, fs::copy_options::overwrite_existing);
+        LogPrintf("copied %s to %s\n", fs::PathToString(m_filepath), fs::PathToString(dst));
+        return true;
+    } catch (const fs::filesystem_error& e) {
+        LogPrintf("error copying %s to %s - %s\n", fs::PathToString(m_filepath), fs::PathToString(dst), fsbridge::get_filesystem_error_message(e));
+        return false;
+    }
```
> Is this used in this pull request?

No, it's used in the followup.

> I wonder if it could make sense to return the error string to the caller, assuming this is called by RPC eventually?

`Backup` is a part of the `WalletDatabase` parent class, so changing this would also mean changing the `Backup` of existing database classes, and all of their callers. I think that's something that would be worth doing in a followup, but not this PR.

achow101 committed:
test: Test bdb_ro dump of wallet without reset LSNs

achow101 committed:
wallet, test: Be able to always swap BDB endianness

laanwj commented on src/wallet/init.cpp:
```diff
@@ -87,8 +87,9 @@ void WalletInit::AddWalletOptions(ArgsManager& argsman) const
+    argsman.AddArg("-swapbdbendian", "Swaps the internal endianness of BDB wallet databases (default: false)", ArgsManager:: ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::WALLET_DEBUG_TEST);
 #else
-    argsman.AddHiddenArgs({"-dblogsize", "-flushwallet", "-privdb"});
+    argsman.AddHiddenArgs({"-dblogsize", "-flushwallet", "-privdb", "swapbdbendian"});
```
CI linter complains, apparently the option needs to be added in another place:
```
AssertionError: Please add {'-swapbdbendian'} to the hidden args in DummyWalletInit::AddWalletOptions
```

laanwj commented:
New tests LGTM
Re-ACK a0943b812ef38826a4ee2732af5f24e470281117

furszy reviewed (commented):
Going slowly here, I want to go through the bdb code diffs too. But, so far, so good. Benchmark 8f2267ff24acdc shows no diff with master and also have created and migrated a v0.6.3 wallet without any problem.

josibake commented:
Concept ACK

Mostly focusing my review on https://github.com/bitcoin/bitcoin/pull/26606/commits/5f0cddb4f66f2a174babdfb17bc31a3011e5cc20. Regarding testing, it occurred to me that we had some large / interesting wallet files from when we were doing wallet simulations for coin selection. @murchandamus @achow101 any chance those wallets were BDB and if so, would be useful to test here?

TheCharlatan commented on src/wallet/migrate.cpp:
```diff
@@ -564,6 +564,24 @@ void BerkeleyRODatabase::Open()
         throw std::runtime_error("BDB builtin encryption is not supported");
     }
 
+    // Check all LSNs point to file 0 and offset 1 which indicates that the LSNs were
```
Nit: Would be nice to say what LSN means, e.g. "Check all LSNs (log sequence numbers)..."

achow101 replied:
Done

sedited reviewed (approved):
Re-ACK a0943b812ef38826a4ee2732af5f24e470281117

The test improvements look good and give some more confidence.

Can you run the commits with newly added files through `contrib/devtools/clang-format-diff`? There are a bunch of simple whitespace issues.

It would also be nice to mention the byte-swapped database type and that it is only used for testing purposes until the final deprecation in the pull request description.

josibake commented on src/wallet/migrate.cpp:
```diff
@@ -1,13 +1,622 @@
+    // Verify the size of the file is a multiple of the page size
+    db_file.seek(0, SEEK_END);
+    int64_t size = db_file.tell();
+    /* BDB doesn't actually error on this
+    if (size % page_size != 0) {
+        throw std::runtime_error("File size is not a multiple of page size");
+    }
+    */
```
in https://github.com/bitcoin/bitcoin/pull/26606/commits/5f0cddb4f66f2a174babdfb17bc31a3011e5cc20 ("wallet: implement independent BDB deserializer in BerkeleyRODatabase"):

nit: I somewhat know whats going here on based on previous discussions about this PR, but without that prior context this being commented out is pretty confusing and I don't think the comment above provides enough explanation. Perhaps add more context in the comment or just remove this snippet entirely?

achow101 replied:
Expanded the comment.

josibake commented on src/wallet/migrate.cpp:
```diff
@@ -750,4 +751,18 @@ std::unique_ptr<DatabaseCursor> BerkeleyROBatch::GetNewPrefixCursor(Span<const s
+        std::unique_ptr<BerkeleyRODatabase> db = std::make_unique<BerkeleyRODatabase>(data_file);
+        status = DatabaseStatus::SUCCESS;
+        return db;
+    } catch (const std::runtime_error& e) {
+        error.original = e.what();
+        status = DatabaseStatus::FAILED_LOAD;
+        return nullptr;
```
in https://github.com/bitcoin/bitcoin/pull/26606/commits/e643cac230fc2ba059847bd4e196bddfd97b1e8e ("Add MakeBerkeleyRODatabase"):

Seems like we should have some sort of call to action here prompting the user to open an issue / reach out on IRC if this fails? If we fail to open a bdb database here, the database itself could be fine and it instead might be an issue with our parser.

achow101 replied:
I don't think that's really necessary? I would expect that users would complain if a wallet suddenly was not able to be opened.

josibake reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/26606/commits/a0943b812ef38826a4ee2732af5f24e470281117

Spent some time comparing this to the bdb header files and LGTM. Test coverage also looks good. Left a few nits regarding comments. Overall, I do think we should prompt users to file an issue specifically for this if loading a database fails, considering it might be nothing wrong with their database and instead an edge case being hit in the parser.

achow101 commented on src/wallet/bdb.cpp:
```diff
@@ -300,7 +300,11 @@ static Span<const std::byte> SpanFromDbt(const SafeDbt& dbt)
 }
 
 BerkeleyDatabase::BerkeleyDatabase(std::shared_ptr<BerkeleyEnvironment> env, fs::path filename, const DatabaseOptions& options) :
```
Done

furszy reviewed (commented):
Despite the CI failures, light ACK abbbb1a38d28bf365fb0f219fc60993e48627631.

achow101 commented on src/wallet/bdb.cpp:
```diff
@@ -389,6 +393,10 @@ void BerkeleyDatabase::Open()
             }
 
+            if (m_byteswap) {
+                pdb_temp->set_lorder(std::endian::native == std::endian::little ? 4321 : 1234);
```
Done

josibake replied:
Looks like this warning is still failing the CI, perhaps just disabling the `unreachable-code` warning here is the simplest?

achow101 commented on src/wallet/test/fuzz/wallet_bdb_parser.cpp:
```diff
@@ -0,0 +1,139 @@
+
+FUZZ_TARGET(wallet_bdb_parser, .init = initialize_wallet_bdb_parser)
+{
+    FuzzedDataProvider fuzzed_data_provider{buffer.data(), buffer.size()};
```
Done

achow101 commented on src/wallet/bdb.cpp:
```diff
@@ -389,6 +393,11 @@ void BerkeleyDatabase::Open()
             }
 
+            if (m_byteswap) {
+                constexpr auto lorder{std::endian::native == std::endian::little ? 4321 : 1234};
```
I think it's fixed this time.

josibake commented:
reACK https://github.com/bitcoin/bitcoin/pull/26606/commits/c0e0a0affea3e3cee9a8910de4bae55a9bdd24cf

sedited reviewed (approved):
Re-ACK c0e0a0affea3e3cee9a8910de4bae55a9bdd24cf

achow101 commented on src/wallet/migrate.cpp:
```diff
@@ -1,13 +1,617 @@
+        uint32_t curr_page = pages.back();
+        if (curr_page > inner_meta.last_page) {
+            std::runtime_error("Page number is greater than subdatabase last page");
+        }
```
Good catch. It looks like BDB doesn't actually set last_pgno to the correct value, so this check is too strict. I've commented it out entirely.

sedited reviewed (approved):
Re-ACK d51fbab4b32d56765e8faab6ad01245fb259b0ca

josibake commented:
reACK https://github.com/bitcoin/bitcoin/commit/d51fbab4b32d56765e8faab6ad01245fb259b0ca

furszy reviewed (commented):
reACK d51fbab4b32d56765e8faab6ad01245fb259b0ca

laanwj reviewed (approved):
re-ACK d51fbab4b32d56765e8faab6ad01245fb259b0ca

theStack reviewed (approved):
ACK d51fbab4b32d56765e8faab6ad01245fb259b0ca

Convinced myself that this parser works reliably partly by looking at the original code, partly by extending the test framework's BDB parser to match this PR's logic (see #30125), which helped a lot in understanding the structure. Also tested with a few random signet legacy wallets that I still had around, though those were all quite small.

Btw, I couldn't manage to find or create a BDB wallet where the DELETE flag is set in a record header (would be nice to exercise that in a test, though it's hard to verify that it's really set without inspecting manually, or e.g. adding debug messages on deserialization), but the code to handle that is simple enough and looks correct.

hebasto commented:
Ported to the CMake-based build system in https://github.com/hebasto/bitcoin/pull/220.

hebasto reviewed (commented):
4d7a3ae78e55f25868979f1bd920857a4aecb825

The new `src/wallet/test/fuzz/wallet_bdb_parser.cpp` [fails](https://github.com/bitcoin/bitcoin/pull/30755#issuecomment-2318990239) on Windows.

achow101 commented:
> The new `src/wallet/test/fuzz/wallet_bdb_parser.cpp` [fails](https://github.com/bitcoin/bitcoin/pull/30755#issuecomment-2318990239) on Windows.

This looks like an inconsistency in BDB itself as the failure is on BDB loading the database file. Additionally, since this test case passes on linux, I'm not entirely sure what to do about it. This will also be resolved once BDB is removed.
